Math/NumberTheory/Moduli/Sqrt.hs

```diff
@@ -13,30 +13,78 @@
 {-# LANGUAGE CPP          #-}
 
 module Math.NumberTheory.Moduli.Sqrt
-  ( sqrtModP
-  , sqrtModPList
-  , sqrtModP'
-  , tonelliShanks
-  , sqrtModPP
-  , sqrtModPPList
+  ( sqrtModList
+  , sqrtModMaybe
+  , sqrtModSharp
   , sqrtModF
   , sqrtModFList
   ) where
 
+import Safe (headMay)
+
 import Control.Monad (liftM2)
 import Data.Bits
 import Data.List (nub)
 import GHC.Integer.GMP.Internals
 
+import Math.NumberTheory.GCD (coprime)
 import Math.NumberTheory.Moduli.Chinese
 import Math.NumberTheory.Moduli.Jacobi
 import Math.NumberTheory.Primes.Sieve (sieveFrom)
+import Math.NumberTheory.Primes.Testing.Probabilistic (isPrime)
+import Math.NumberTheory.Primes.Factorisation.Montgomery (factorise)
 import Math.NumberTheory.Utils (shiftToOddCount, splitOff)
 
+-- | Extended flag structure to check if square root is computed in finite field.
+data FieldCharacterictic = Invalid
+                         | Prime
+                         | PrimePower Integer Int
+  deriving Show
+
+-- | Check if value is power of some prime.
+checkPrimePower :: Integer -> Maybe (Integer, Int)
+checkPrimePower = headMay . factorise
+
+-- | Check if argument is valid characteristic of finite field.
+--   It's a precondition for computations in functions like @'sqrtModP'@ and @'sqrtModPP'@
+checkPreconditions :: Integer -> FieldCharacterictic
+checkPreconditions p
+  | isPrime p = Prime
+  | otherwise = case checkPrimePower p of
+      Nothing -> Invalid
+      Just (prime,pow) -> PrimePower prime pow
+
+-- * Interface functions with preconditions checking.
+
+sqrtModSharp :: Integer -> Integer -> Integer
+sqrtModSharp n p
+  | p <= 0 = error "Cannot extract sqaure root by negative modulo"
```

Comment:
`sqaure` - typo. 
`negative` - should be `non-positive`, because zero is not permitted also.